Multi process model framework - Example module in test folder

mak-454 opened the pull request:
Multi process framework for service controller.
Attached is the sample unit test case module.

mak-454 committed:
Multi process model framework - Example module in test folder

mak-454 commented:
@mageshgv-oc @songole @chandra-k @njagadish

mageshgv-oc commented:
@hemanthravi @ashutosh-mishra @rajendramachani

mageshgv-oc commented:
Ahmed, there are a bunch of pep8 errors. can you please fix them.
./run_tests.sh -p

mak-454 commented:
where is this ./run_tests.sh ?

mageshgv-oc commented:
It is part of all openstack projects framework. Same as the details sent in the mail on unit tests.
https://wiki.openstack.org/wiki/Testing
http://docs.openstack.org/developer/hacking/

mak-454 committed:
Fixed bunch of pep8 errors

mak-454 committed:
Fixed multiple rpc connection issue

mak-454 committed:
RpcCallback is not supported from kilo - corrected it, fixed some other import errors in test module

mageshgv-oc commented on gbpservice/neutron/nsf/core/test/visibility.py:
```diff
@@ -0,0 +1,175 @@
+        # Filebeat(**self._service).delete_configure()
+
+
+class RpcManager(n_rpc.RpcCallback):
```
RpcCallback class is not present in the liberty Neutron.

mak-454 replied:
Corrected it !

mak-454 committed:
Added modules folder, corrected test module for liberty

mak-454 committed:
Corrected the imports as per liberty release

mak-454 committed:
Corrected the log message in test module to fix syntax error issue

mak-454 committed:
Implemented rest client over unix socket, unix to tcp proxy

DhuldevValekar committed:
Devstack changes to run process model(NSF).

mak-454 committed:
Implemented max times for timer events, support for decorated functions with variable spacing and event handlers, proxy for unix2tcp

mak-454 committed:
Merge branch 'multi_process_framework' of https://github.com/oneconvergence/group-based-policy into multi_process_framework

mak-454 commented:
As came out in the last review meeting -- a) Added support to provide max_times for a poll event, an event will be polled this many times at max after that event is removed from pollQ and if registered handler implements "poll_cancel" method it is invoked. b) Added support to be able to write decorated methods with different spacing for each event, if a particular event handler does not implement any decorated function then the default "handle_poll_event" will be invoked. This allows in openstack style to provide different timeouts for different events.

mak-454 committed:
Fixed the issues observed while unit testing proxy

mak-454 committed:
1) Implemented queue class which can be used by configurator to put the rpccallbacks, core also uses it for caching 2) Added support for invoking init_completed method - modules can implement this method if they wish to add some events immediately after init is completed

mak-454 committed:
1) Fixed a issue in module loading logic 2) Implemented more scalable proxy handling - proxy connections are completely handled asynchronously and is shared across threads - this way more num of connections can be handled

mak-454 committed:
Added test code for proxy

mak-454 committed:
Added support for agent report_state in framework

mak-454 committed:
Removed pdb, corrected a bug in periodic task impl, moved logging setup before module imports

njagadish commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,582 @@
+        for module in modules:
+            try:
+                module.module_init(self, self._conf)
+            except AttributeError as s:
```
Why only AttributeError is caught. Modules can fail for various reasons right?

mak-454 replied:
At that line we are trying to access an attribute of a class -- so only possibility is if that attribute is not present. what other exceptions do you expect ?

njagadish replied:
Since it is calling module_init, what ever functionality modules add, based on that exceptions can change and they can be raised in the modules.

mak-454 committed:
a) Removed unused code b) Code cleanup c) Added proper logs

mak-454 committed:
Updated the logs to right levels

AkashDeepSrivastava committed:
included periodic interval and reportstate_interval as configurable data

mak-454 committed:
Corrected the thread name identify logic

mak-454 committed:
Merge branch 'multi_process_framework' of https://github.com/oneconvergence/group-based-policy into multi_process_framework

hemanthravi commented on gbpservice/neutron/config/nsf.ini:
```diff
@@ -0,0 +1,13 @@
+rabbit_userid=stackrabbit
+rabbit_hosts=<openstack controller ip adress:rabbitPort>  #Example 192.168.2.107:5672
+rabbit_port=5672
+rabbit_host=<openstack controller ip address>
```
Can these be derived from existing ini files?
rabiit_hosts contain the same info as rabbit_port, rabbit_host.

mak-454 replied:
Yes those will be derived from neutron.conf, they were put in separate .ini file to highlight what parameters the framework need. Finally the framework run as /usr/bin/python /usr/bin/nfp --config-file /etc/neutron/neutron.conf --config-file /etc/neutron/nfp.ini

mak-454 replied:
Ok will comment out these params and checkin the new file.

mak-454 replied:
For configurator we might not need /etc/neutron/neutron.conf -- so this file /etc/nsf/nsf.ini will be useful.. Even inside configurator vm, we decided to use rabbitmq for the communication between rest server and configurator. So, retaining this file for now.

mak-454 committed:
Commented rmq options from nsf.ini - these can be picked up from /etc/neutron/neutron.conf

songole commented on gbpservice/neutron/nsf/core/cfg.py:
```diff
@@ -0,0 +1,35 @@
+        help=_('#of workers to create.')
+    ),
+    cfg.StrOpt(
+        'RpcLoadBalancer',
```
Can we follow a consistent model. all lower case or all upper case?

Why is this called RpcLoadBalancer? Is it distributing across queues?

mak-454 replied:
Actually the idea was if we could implement any good loadbalancing algorithm for distribution of events to worker process, we should be able to select it from this config options v/s the code change in framework. New loadbalancer algo can be implemented in core/lb.py and this conf option is used to make code point to that loadbalancer. Ok will make its name consistent and change its name to "event_loadbalancer"

mak-454 replied:
Done. Please verify in latest commit.

songole commented on gbpservice/neutron/nsf/core/cfg.py:
```diff
@@ -0,0 +1,35 @@
+    cfg.IntOpt(
+        'evs_polling_interval',
+        default=1,
+        help=_('Polling interval for events in seconds ')
```
Can we add a little more help for all the parameters. Docs will be generated based on these help text. Lets make sure they make sense.

mak-454 replied:
a) Ok will add more help text
b) 'evs_polling_interval' is unused will remove it.
c) 'polling_interval' -- Is the minimum interval with which periodic task must poll. Events could have different spacing relative to this polling interval.
d) 'report_state_interval' -- This is the interval at which neutron type of *aaS agents should report their state to their corresponding plugins.

mak-454 replied:
Done. Please verify if help text is approp now. Also removed unused params.

songole commented on gbpservice/neutron/nsf/core/lb.py:
```diff
@@ -0,0 +1,46 @@
+
+    def get(self, rsrcid):
+        if rsrcid in self._assoc.keys():
+            worker = self._assoc[rsrcid]
```
Can we not use a hash on src id? Are we worried about skewing?

mak-454 replied:
"dict" is sort of hash right ? Here, in the logic we are trying to get the bound worker for the given resource identified by rsrc_id.

songole commented on gbpservice/neutron/nsf/core/lb.py:
```diff
@@ -0,0 +1,46 @@
+        return self._rr()
+
+    def rem(self, rsrcid):
+        pass
```
Who calls rem?

mak-454 replied:
Actually no one.. before writing the logic, had put template for the class with "put,get,rem" etc methods.. will remove it.

mak-454 replied:
Removed this unused methods.

songole commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+    except:
+        return ""
+
+
```
Can we add some comments? I don't see any in any of the files..

mak-454 replied:
OK.

mak-454 replied:
Added comments across file for functions and classes. Please check in latest commit.

songole commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        return "{0}.{1}.{2}".format(
+            type(obj.im_self).__module__,
+            type(obj.im_self).__name__,
+            obj.__name__)
```
What is the difference between im_self.__name__ and __name__?

mak-454 replied:
im_self.name gives the class name
name gives the method name.
So, finally the display would be like <module>.<class>.<method>

songole commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+from oslo_service import periodic_task as oslo_periodic_task
+from oslo_service import loopingcall
+from oslo_service import service as os_service
+
```
It would be confusing to call some as oslo_, some as os_ and some others without any prefix etc. Can we call it oslo_service instead of os_service?

mak-454 replied:
true. Will make it consistent.

mak-454 replied:
Made it consistent. Please check latest commit.

songole commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+
+
+class RpcAgents(object):
+
```
Can we get an idea of how many RPC agents and what they are?

mak-454 replied:
Actually, the process does process RPC calls.. i mean the RPCAgents are invoked from main process and registered to it.
Config Agent has --- a) VPN b) FW c) LB d) SM
Similarly orchestrator and configurator also has RPC endpoints.

songole commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        for agent in self.services:
+            agent.report_state()
+
+
```
I think we need comments. What is agent reporting and what is the following periodic reporting?

mak-454 replied:
Will add comments.
Agent reporting is periodic reporting --- Actually some of the RPC Endpoints we have are neutron *aaS agents -- one requirement for *aas agent is to report its state periodically. In our design each agent could be a nsf-module or multiple agents can be in single module, so if it was allowed for each one to write their own logic it was little difficult. since this was common requirement for most of the code, state reporting is incorporated in the framework- to handle it in single common place to make it less error prone.
To report state like this, we need a periodic task -- hence another periodic task is started to handle report stating and there is one more periodic task which handles the event polling.

mak-454 replied:
Added comments in latest commit.

hemanthravi commented on gbpservice/neutron/nsf/core/lb.py:
```diff
@@ -0,0 +1,46 @@
+        return worker
+
+    def rem(self, rsrcid):
+        del self._assoc[rsrcid]
```
Rename file lb.py - to avoid any confusion with lb service.
Make method and var names be descriptive - This is a comment across all files.

mak-454 replied:
a) Will change the file name.
b) Yes will add comments across the files.

mak-454 replied:
Done.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+            LOG.exception(_("Stopped reporting agent state!"))
+
+
+class RpcAgent(n_rpc.Service):
```
Does this make it a neutron agent that shows up in neutron agent-list? Is this intended?

mak-454 replied:
Yes, this code is required for neutron agent shows up in the neutron agent-list.
The way you are suggesting is difficult to implement.
There is only a single module, but not single agent, it will register multiple agents and multiple RpcEndpoints for better handling. And each agent like FW, VPN, LB has different data to report to the plugin. Please look at "config_agent_modules" git branch and see modules/config_agent.py

mak-454 replied:
With neutron there is no way to register single agent with multiple topics. One agent one topic. It has to be multiple rpc agents.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        self._sc = sc
+        self._cache = core_queue.Queue(sc)
+
+    def rem(self, ev):
```
remove instead of rem

mak-454 replied:
OK.

mak-454 replied:
This class looked redundant. Removed it.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        return self._cache.copy()
+
+
+class Serializer(object):
```
Rename to EventSerializer?
As OSM suggested, use a consistent scheme '_' separated lower case or Caps. What is the scheme used in GBP code?

mak-454 replied:
a) Will talk to magesh to know the GBP scheme
b) Ok, will change the name.

mak-454 replied:
There is no such scheme in GBP itseems.
Changed the name to Sequencer

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+                "There is already an event in progress"
+                "Queueing event %s" % (ev.identify())))
+
+            mapp[ev.binding_key]['queue'].append(ev)
```
Queued here since it can't be put on the event_queue. Is that correct?
Where is the code to dequeue this event and add it to the event_queue?

mak-454 replied:
Yes, it is stored in the temp cache because some other related event is already being processed.
Please check, "class EventHandler --> _get()" method.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        self._sc.lock()
+        if ev.worker_attached not in self._serializer_map:
+            self._serializer_map[ev.worker_attached] = {}
+        mapp = self._serializer_map[ev.worker_attached]
```
mapp - can the name be more descriptive?

mak-454 replied:
Ok.

mak-454 replied:
Changed it to seq_map

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        self._sc.unlock()
+
+
+class PollHandler(object):
```
Comments in this class are suggestions for code readability. Use what you feel are appropriate.

PollQueue instead of PollHandler

mak-454 replied:
OK.

mak-454 replied:
Added comments across file. Using PollQueueHandler.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        LOG.debug(_("Add event %s to the pollq" % (event.identify())))
+        self._pollq.put(event)
+
+    def rem(self, event):
```
remove

mak-454 replied:
OK.

mak-454 replied:
Updated.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        pull = qlen if (idx + count) > qlen else count
+        return cache[idx:(idx + pull)], pull
+
+    def event_done(self, ev):
```
Remove this method. Call self.rem() where event_done is invoked.

mak-454 replied:
OK.

mak-454 replied:
Actually this is consistent with logic in other places, *_event_done is called to handle cleanup of the event when it is complete. Retaining it.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+            return ev
+        return None
+
+    def event(self, ev):
```
handle_event instead of event

mak-454 replied:
OK.

mak-454 replied:
Removed the "event" method and added logic in "_process_event" itself. Please check the latest commit.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        LOG.debug(_("Processing poll event %s" % (ev.identify())))
+        self.event_done(ev) if ev.id == 'POLL_EVENT_DONE' else self.event(ev)
+
+    def poll(self):
```
run instead of poll

mak-454 replied:
OK.

mak-454 replied:
Done.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+                ev.max_times -= 1
+                ev.last_run = ev1.last_run
+
+    def process(self, ev):
```
Remove this method. Instead caller invokes the contents of this method.

mak-454 replied:
Actually it will be better to do this in separate method.

mak-454 replied:
Removed it.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+class EventHandlers(object):
+
+    def __init__(self):
+        self._ehs = {}
```
_event_handlers instead of _ehs?

mak-454 replied:
OK

mak-454 replied:
Class name and variable name will be confusing. Since class name is already giving the information, retaining the variable as is.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        self._lock.release()
+        LOG.debug(_("Released lock.."))
+
+    def event_done(self, ev):
```
event_done....not descriptive. Is this meant to be called after the event is handled?
How about cleanup_event?

mak-454 replied:
This means that event is complete all the required cleanup needs and any other actions needs to be handled. Does "event_complete" makes more sense ? Or a descriptive comment.

mak-454 replied:
Added description as comment.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+
+    def event_done(self, ev):
+        LOG.info(_("Event %s done" % (ev.identify())))
+        mapp = self._serializer.copy()
```
rename mapp

mak-454 replied:
OK.

mak-454 replied:
Done.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+
+        LOG.debug(_("Checking if event %s in serialize Q"
+                    % (ev.identify())))
+        qu = mapp[ev.binding_key]['queue']
```
queue instead of qu

mak-454 replied:
thought it might be confusing.. because we are using "queue" as class. Since it is a local variable, can we retain it ?

mak-454 replied:
modified it to seq_q ( sequencer_queue)

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        '''
+        self._serializer_map = {}
+
+    def serialize(self, ev):
```
Rename to schedule?

mak-454 replied:
But this is not scheduling.. it is serializing..

mak-454 replied:
Class name is changed to sequencer.
Methods are changed to "add", "remove"

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        self._sc.unlock()
+        return queued
+
+    def deserialize(self, ev):
```
Is this supposed to be cleanup? Rename to cleanup?

mak-454 replied:
Not its not cleanup.

mak-454 replied:
Changed the method name to 'remove'. It is just removing the completed entry from the queue of serial map.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+                break
+        self._serializer.deserialize(ev)
+
+    def serialize(self, ev):
```
event instead of ev? queue instead of qu? Suggestion only.

hemanthravi replied:
schedule instead of serialize?

mak-454 replied:
We have classes with those names.. so not to get confused named them like that. Since there are local variables, should be ok I guess.

mak-454 replied:
updated 'ev' to 'event'. Changed 'qu' to 'seq_q'

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+            return ev
+        return None
+
+    def serialize_get(self):
```
get_pending_event.....get_queued_event....get_next_event ?

mak-454 replied:
a) Ok. will use 'get_next_event'

mak-454 replied:
modified it to 'sequencer_put_event' & 'sequencer_get_event'

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+
+        for w in range(0, wc):
+            qu = Queue()
+            evworker = EventHandler(self, qu, self.ehs)
```
ev_worker or event_worker instead of evworker?

hemanthravi replied:
evworker should be event_quue_handler. This is not the worker process.

mak-454 replied:
a) will change evworker to ev_queue_handler

mak-454 replied:
Done.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+            evworker = EventHandler(self, qu, self.ehs)
+            proc = Process(target=evworker.run, args=(qu,))
+            proc.daemon = True
+            workers[w] = workers[w] + (proc, qu, evworker)
```
ev_workers instead of workers?

hemanthravi replied:
Should the list (proc, qu, evworker) be a dict?
workers should be event_worker_descs

mak-454 replied:
a) OK will change 'workers' to 'ev_workers'
b) It need not be a 'dict' we are never key'ing in to it. We just loop over it.

mak-454 replied:
Done.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        for w in range(0, wc):
+            qu = Queue()
+            evworker = EventHandler(self, qu, self.ehs)
+            proc = Process(target=evworker.run, args=(qu,))
```
How are the following handled:
- main process is killed
- any worker process is killed

hemanthravi replied:
proc should be the event_worker or event_worker_proc

mak-454 replied:
a) We had collectively decided to put this for later --- Process getting killed. We had multiple discussions on this whether to replace Queue to "redisQ" for persistence and read from it after restart.
b) lets retain proc.

mak-454 replied:
modified 'proc' to 'worker'

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        event.max_times = max_times
+        self.pollhandler.add(event)
+
+    def poll_event_done(self, event):
```
Can this be the same as poll_event? Caller sets event.id = ..DONE and calls with max_times = 1?

mak-454 replied:
Lets retain the 'poll_event_done' as there is 'event_done' method as event completion methods.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+    def register_rpc_agents(self, agents):
+        self.rpc_agents.add(agents)
+
+    def event(self, **kwargs):
```
create_event ?

mak-454 replied:
may be new_event.

mak-454 replied:
Modified it to 'new_event' -- using 'create_event' for 'rpc_event'

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+        self._procidx = (self._procidx + count) % (self._batch)
+
+
+class EventHandler(object):
```
EventQueueHandler instead of EventHandler?

mak-454 replied:
OK

mak-454 replied:
Done.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+    sc.start()
+    sc.init_complete()
+    # sc.unit_test()
+    sc.wait()
```
Need comments through out this file.
Is this file too big? Is there a coding guideline for file size? If this is true can it be broken up?

mak-454 replied:
a) Will figure out about the coding guideline about file size from magesh.
b) Yes will add the comments.

mak-454 replied:
a) There is no such guide line for file size it seems.
b) Added comments.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+    def __init__(self):
+        self._ehs = {}
+
+    def register(self, ev):
```
Rename 'ev' to something like event_desc or event_info, to differentiate from an event in the queue.

mak-454 replied:
'ev' is here the 'event' in the queue.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+
+class EventHandler(object):
+
+    def __init__(self, sc, qu, eh):
```
Using 'eh' and '_eh' below makes this hard to read. This class is EventHandler (eh) and you are actually passing in a event_desc_table. As noted above EventHandler should be renamed since it's processing the event_queue. eh, _eh are actually event_desc_tables?

mak-454 replied:
OK. Will make approp changes.

mak-454 replied:
EventHandler is renamed to EventQueueHandler
_eh is changed to _ehs -- Actually this is object of class 'EventHandlers' -- This class is cache of all the registered event handlers.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+            time.sleep(0)  # Yield the CPU
+
+
+class EventHandlers(object):
```
This is not a table of the prev class EventHandler. This is more a table of event_desc each of which describes how an event should be handled. Rename to EventDescTable?

mak-454 replied:
OK.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+class EventHandler(object):
+
+    def __init__(self, sc, qu, eh):
+        self._tpool = core_tp.ThreadPool()
```
Does this threadpool get attached to the one of the worker processes?

mak-454 replied:
Every process a pool of threads.

hemanthravi commented on gbpservice/neutron/nsf/core/queue.py:
```diff
@@ -0,0 +1,74 @@
+from collections import deque
+
+
+class Queue(object):
```
Rename to avoid confusion with multiprocessing Queue ?

mak-454 replied:
OK, to Fifo() - because that is what this queue is.

mak-454 replied:
Done.

hemanthravi commented on gbpservice/neutron/nsf/core/queue.py:
```diff
@@ -0,0 +1,74 @@
+    def _qsize(self):
+        return len(self._q)
+
+    def _empty(self):
```
rename to _is_empty?

mak-454 replied:
OK

hemanthravi commented on gbpservice/neutron/nsf/core/queue.py:
```diff
@@ -0,0 +1,74 @@
+            self._sc.lock()
+            self._empty()
+            for i in range(0, limit):
+                items = self._pop(items)
```
What happens if q has less than limit items?

mak-454 replied:
Only the available number of items are returned. Plz see, Queue.Empty exception is handled.

hemanthravi commented on gbpservice/neutron/nsf/core/queue.py:
```diff
@@ -0,0 +1,74 @@
+            self._sc.unlock()
+
+    def get(self, limit=sys.maxint):
+        items = []
```
msgs instead of items to be consistent with put?

mak-454 replied:
OK.

mak-454 replied:
Done

hemanthravi commented on gbpservice/neutron/nsf/core/queue.py:
```diff
@@ -0,0 +1,74 @@
+            self._sc.unlock()
+            return items
+
+    def copy(self):
```
Does this make a copy of all the items in the q?

mak-454 replied:
Right.. in case we do not want to pop the element from Queue.

hemanthravi commented on gbpservice/neutron/nsf/core/queue.py:
```diff
@@ -0,0 +1,74 @@
+        try:
+            self._sc.lock()
+            for item in items:
+                self._q.remove(item)
```
What happens if item doesn't exist in q?

mak-454 replied:
It will throw 'indexerror' exception, caller has to handle it.

hemanthravi commented on gbpservice/neutron/nsf/core/periodic_task.py:
```diff
@@ -0,0 +1,106 @@
+import random
```
Need a walk through of this code.

hemanthravi commented on gbpservice/neutron/nsf/core/main.py:
```diff
@@ -0,0 +1,705 @@
+from neutron import context
+from neutron.agent import rpc as agent_rpc
+
+from gbpservice.neutron.nsf.core import cfg as core_cfg
```
nsf everywhere should be nfp
Instead of core_xxx, use nfp_xxx. Will the be more readable?

mak-454 replied:
OK. will change.

mak-454 replied:
Done.

hemanthravi commented on gbpservice/neutron/nsf/core/threadpool.py:
```diff
@@ -0,0 +1,80 @@
+
+    def __init__(self, thread, pool):
+        self.thread = thread
+        self.thread.link(_thread_done, pool=pool, thread=self)
```
What does link do?

mak-454 replied:
It calls the _thread_done() method whenever the thread stops.. in case we have to handle anything.. we are removing the thread object from our cache.

hemanthravi commented on gbpservice/neutron/nsf/core/threadpool.py:
```diff
@@ -0,0 +1,80 @@
+LOG = logging.getLogger(__name__)
+
+
+def _thread_done(gt, *args, **kwargs):
```
When is _thread_done invoked?

mak-454 replied:
When thread is stopped.

hemanthravi commented on gbpservice/neutron/nsf/core/threadpool.py:
```diff
@@ -0,0 +1,80 @@
+        self.thread = thread
+        self.thread.link(_thread_done, pool=pool, thread=self)
+
+    def stop(self):
```
Who invokes stop, wait, link?

mak-454 replied:
these are available for callers to invoke. Right now there is no condition where we want to stop thread in between. It finishes completely.

hemanthravi commented on gbpservice/neutron/nsf/core/threadpool.py:
```diff
@@ -0,0 +1,80 @@
+    def thread_done(self, thread):
+        self.threads.remove(thread)
+
+    def stop(self):
```
Who invokes stop, wait?

mak-454 replied:
these are available for callers to invoke. Right now there is no condition where we want to stop thread in between. It finishes completely.

hemanthravi commented on gbpservice/neutron/nsf/core/threadpool.py:
```diff
@@ -0,0 +1,80 @@
+                # don't kill the current thread.
+                continue
+            try:
+                x.stop()
```
Will this call result in calling thread_done and hence cause the list of threads to be modified? Is this an issue? Does this mismatch with the comment above?

mak-454 replied:
Right.

mak-454 committed:
Addressed all the review comments

hemanthravi commented on gbpservice/neutron/nsf/proxy/proxy.py:
```diff
@@ -0,0 +1,223 @@
+    def recv(self):
+        self._socket.settimeout(self._idle_wait)
+        try:
+            data = self._socket.recv(16)
```
Can the remote side close the connection? How do we handle errors from recv()?

hemanthravi replied:
Does recv() block until timeout?
Is it possible to use a select() like mechanism?

mak-454 replied:
With select also we would need a timeout. For some reason python select is not working like 'c' select. it is blocking the context.

hemanthravi commented on gbpservice/neutron/nsf/proxy/proxy.py:
```diff
@@ -0,0 +1,223 @@
+        return None
+
+    def send(self, data):
+        self._socket.send(data)
```
Can the remote side close the connection? How do we handle errors from recv()?

mak-454 replied:
Yes remote side can close connection. All send exceptions are catch all. Please see line 156 (run) method.
w.r.t 'recv' the connection in python was not coming out if remote side closes the connection in certain conditions, so had to handle timeout socket error and also put some explicit timeout on read operation.

hemanthravi commented on gbpservice/neutron/nsf/proxy/proxy.py:
```diff
@@ -0,0 +1,223 @@
+        self._tcp_conn.close()
+
+    def _proxy(self, rxconn, txconn):
+        data = rxconn.recv()
```
Should we read until there is more data to read?

mak-454 replied:
We cannot because if one particular socket always has data then that task will always work blocking the other threads and other connections. To be fair some timeout must be associated.

hemanthravi commented on gbpservice/neutron/nsf/proxy/proxy.py:
```diff
@@ -0,0 +1,223 @@
+class Worker(object):
+
+    def run(self):
+        while True:
```
Will this loop use 100% CPU?

mak-454 replied:
No it will not, as we are scheduling out this thread everytime using time.sleep(0)

hemanthravi commented on gbpservice/neutron/nsf/proxy/proxy.py:
```diff
@@ -0,0 +1,223 @@
+
+    def start(self):
+        for i in range(self.conf.worker_threads):
+            t = threading.Thread(target=Worker().run)
```
Why Thread here and Process in the nsf core?

mak-454 replied:
We may not need multiple process here - because cpu bound logic is less.. mostly it is I/O so threads should be sufficient to handle multiple connections.

mak-454 committed:
Addressed the review comments

mak-454 committed:
Addressed the review comment of splitting one main file into multiple files, event handling into one, rpc into one, polling into one

Rahul2702 committed:
bug-1234, Added comments and test cases for proxy

mak-454 committed:
Fixed the issue of job distribution to different workers when binding key is not specified

mak-454 committed:
Added license banners

mak-454 committed:
Ran ./run_tests.sh -N --pep8 and fixed the errors reported

mak-454 committed:
As per the review comments, moved code to new folder nfp

mak-454 committed:
Fixed all the errors reported by tox env of gbp repo

mak-454 committed:
Bug: Poll handler is invoked after stopping.

Corner case : After an event timedout after 'N' secs,
module poll handler is invoked, if it takes >N secs
which makes the event timeout one extra time before
module declares it to be complete.
This one extra time poll event would get delivered
to worker and module does not know how to handle it.

Fix: Event is not polled again unless module logic
decides so. The fate of poll event is decided by the
return value of poll handler. Poll timeout could get
shift few ticks, but corner cases wont occur.

Rahul2702 committed:
Removed typo error and tested

Rahul2702 committed:
Test case for poll_event_cancel

Rahul2702 committed:
Added code for Event lifetime Timer

Rahul2702 committed:
Added core_debug parmeter in to cfg to control on debug logs of main module

Rahul2702 committed:
Removed unused code

mak-454 committed:
Fixed the event serialization issues, also added core_debug option to enable/disable core module debug options separately

mak-454 committed:
Merged stashq fixes from integration branch

mak-454 committed:
Fixed tox pep8 errors

mak-454 commented:
Pulled into base branch through different PR. Closing this one.
